src/Raven.Server/Documents/Indexes/Static/TimeSeries/MapTimeSeriesIndex.cs

```diff
@@ -0,0 +1,219 @@
+        {
+            if (isReference)
+                return base.GetLastTombstoneEtagInCollection(databaseContext, collection, isReference);
+
+            // we do not process tombstones for timeseries, just for references
+            return 0;
+        }
+
+        protected override IndexItem GetItemByEtag(DocumentsOperationContext databaseContext, long etag)
+        {
+            var timeSeries = DocumentDatabase.DocumentsStorage.TimeSeriesStorage.GetTimeSeries(databaseContext, etag);
+            if (timeSeries == null)
+                return default;
+
+            return new IndexItem(timeSeries.Key, timeSeries.Key, timeSeries.Etag, default, timeSeries.SegmentSize, timeSeries);
+        }
+
+        protected override IndexItem GetTombstoneByEtag(DocumentsOperationContext databaseContext, long etag)
+        {
+            throw new NotSupportedException("We do not process tombstones for TimeSeries");
+        }
+
+        protected override bool HasTombstonesWithEtagGreaterThanStartAndLowerThanOrEqualToEnd(DocumentsOperationContext databaseContext, IIndexCollection collection, long start, long end)
+        {
+            throw new NotSupportedException("We do not process tombstones for TimeSeries");
+        }
+
+        protected override bool IsStale(DocumentsOperationContext databaseContext, TransactionOperationContext indexContext, long? cutoff = null, long? referenceCutoff = null, List<string> stalenessReasons = null)
+        {
+            var isStale = base.IsStale(databaseContext, indexContext, cutoff, referenceCutoff, stalenessReasons);
+            if (isStale && stalenessReasons == null || _referencedCollections.Count == 0)
+                return isStale;
+
+            return StaticIndexHelper.IsStaleDueToReferences(this, databaseContext, indexContext, referenceCutoff, stalenessReasons) || isStale;
+        }
+
+        internal override IEnumerable<IIndexCollection> GetCollectionsForIndexing()
+        {
+            return _compiled.Maps.Keys;
+        }
+
+        public override Dictionary<IIndexCollection, HashSet<CollectionName>> GetReferencedCollections()
+        {
+            return _compiled.ReferencedCollections;
+        }
+
+        protected override unsafe long CalculateIndexEtag(DocumentsOperationContext documentsContext, TransactionOperationContext indexContext,
+            QueryMetadata query, bool isStale)
+        {
+            if (_referencedCollections.Count == 0)
+                return base.CalculateIndexEtag(documentsContext, indexContext, query, isStale);
+
+            var minLength = MinimumSizeForCalculateIndexEtagLength(query);
+            var length = minLength +
+                         sizeof(long) * 2 * (Collections.Count * _referencedCollections.Count); // last referenced collection etags and last processed reference collection etags
+
```

Comment:
Can users give a LOT of collections here and smash the stack?

Reply:
https://issues.hibernatingrhinos.com/issue/RavenDB-14389